src/python/WMComponent/RucioInjector/RucioInjectorPoller.py

```diff
@@ -370,15 +370,84 @@ def deleteBlocks(self):
+        for blockName in blockDict:
+            container = blockDict[blockName]['dataset']
+            # If the container is not in the dictionary, create a new entry for it
+            if container not in containerDict:
+                # Set of sites to which the container needs to be transferred
+                sites = blockDict[blockName]['sites']
+                containerDict[container] = {'blocks': [], 'rse': sites}
+            containerDict[container]['blocks'].append(blockName)
+
+        for contName in containerDict:
+            cont = containerDict[contName]
+
+            # Checks if the container is not requested in any sites.
+            # This should never be triggered, but better safe than sorry
+            if not cont['rse']:
+                logging.warning("No rules for container: %s. Its blocks won't be deleted.", contName)
+                continue
+
+            try:
+                # Get RSE in which each block is available
+                availableRSEs = self.rucio.getReplicaInfoForBlocks(block=cont['blocks'])
+            except Exception as exc:
+                msg = "Failed to get replica info for blocks in container: %s.\n" % contName
+                msg += "Will retry again in the next cycle. Error: %s" % str(exc)
+                logging.error(msg)
+                continue
+
+            for blockRSEs in availableRSEs:
+                # If block is available at every RSE its container needs to be transferred, the block can be deleted
+                blockSites = set(blockRSEs['replica'])
+                if cont['rse'].issubset(blockSites) or len(blockSites & cont['rse']) >= 2:
+                    blocksToDelete.append(blockRSEs['name'])
+
+        # Delete agent created rules locking the block
+        binds = []
+        logging.info("Going to delete %d blocks", len(blocksToDelete))
+        for block in blocksToDelete:
+            rules = self.rucio.listDataRules(block, scope=self.scope, account=self.rucioAcct)
+            deletedRules = 0
+            if not rules:
+                logging.info("Block rule for: %s has been deleted by previous cycles", block)
+                binds.append({'DELETED': 1, 'BLOCKNAME': block})
+                continue
+
+            for rule in rules:
+                if self.rucio.deleteRule(rule['id']):
+                    logging.info("Successfully deleted rule: %s, for block %s.", rule['id'], block)
+                    deletedRules += 1
+                else:
+                    logging.warning("Failed to delete rule: %s, for block %s.", rule['id'], block)
+
+            if deletedRules == len(rules):
+                binds.append({'DELETED': 1, 'BLOCKNAME': block})
+                logging.info("Successfully deleted every rule for block %s.", block)
+            else:
+                logging.warning("Some rules for block %s were not deleted. Will retry in the next cycle.", block)
+
+        self.markBlocksDeleted.execute(binds)
```

Comment:
Need whether `binds` is empty or not

Reply:
maybe add it to the class carrying the database operation.

Reply:
I added a protection MarkBlocksDeleted.py